internal/jobs/common.go

```diff
@@ -0,0 +1,72 @@
+package jobs
+
+import (
+	"context"
+
+	"github.com/RHEnVision/provisioning-backend/internal/ctxval"
+	"github.com/RHEnVision/provisioning-backend/internal/dao"
+)
+
+func finishWithSuccess(ctx context.Context, reservationId int64) {
+	ctxLogger := ctxval.Logger(ctx).With().Int64("reservation_id", reservationId).Logger()
+
+	rDao, err := dao.GetReservationDao(ctx)
+	if err != nil {
+		ctxLogger.Warn().Err(err).Msg("unable to update job status: dao")
+		return
+	}
+
+	reservation, err := rDao.GetById(ctx, reservationId)
+	if err != nil {
+		ctxLogger.Warn().Err(err).Msg("unable to update job status: get by id")
+		return
+	}
+	ctxLogger.Debug().Msgf("Job step: %d/%d", reservation.Step, reservation.Steps)
+
+	// if this was the last step, set the success flag
+	if reservation.Step >= reservation.Steps {
+		ctxLogger.Info().Msgf("All jobs executed, marking job as success")
+		err = rDao.FinishWithSuccess(ctx, reservationId)
+		if err != nil {
+			ctxLogger.Warn().Err(err).Msg("unable to update job status: finish")
+		}
+	}
+}
+
+func finishWithError(ctx context.Context, reservationId int64, jobError error) {
+	ctxLogger := ctxval.Logger(ctx).With().Int64("reservation_id", reservationId).Logger()
+	ctxLogger.Warn().Err(jobError).Msgf("Job returned an error: %s", jobError.Error())
+
+	rDao, err := dao.GetReservationDao(ctx)
+	if err != nil {
+		ctxLogger.Warn().Err(err).Msg("unable to update job status: dao")
+		return
+	}
+	err = rDao.FinishWithError(ctx, reservationId, jobError.Error())
+	if err != nil {
+		ctxLogger.Warn().Err(err).Msg("unable to update job status: finish")
+	}
+}
+
+func updateStatusBefore(ctx context.Context, id int64, status string) {
+	updateStatusAfter(ctx, id, status, 0)
+}
```

Comment:
I dont understand why did you choose this implementation, can you explain? 😄

Reply:
So one of the drawback of the old design was that Status (string) field was used both for status messages before and after ("something started" as well as "something is done" or "something failed").

Now, the DAO exposes function called `UpdateStatus` that previously only contained `status string` argument, now also contains `addSteps int32` - a relative change in steps that needs to be performed when setting new status.

In the status updates at job start, we do not want to update step counter, so this is when you pass zero. When job ends (regardless if there was error or not - that's what `defer` does), we actually want to increase step by one. So that's why I have two different function, so it is easy to understand for someone who is implementing new job by copying code.

Otherwise the code would need to look something like this, which is pretty not intuitive:

```go
rDao.UpdateStatus(ctx, reservationId, "Starting xyz", 0)

// do some work

rDao.UpdateStatus(ctx, reservationId, "Finished xyz", 1)
```

I could design it via two separate functions, but then these would be two separate SQL queries which is slow:

```go
rDao.UpdateStatus(ctx, reservationId, "Starting xyz")

// do some work

rDao.UpdateStatus(ctx, reservationId, "Finished xyz")
rDao.IncrementStep(ctx, reservationId)
```

But if you like this more, an extra SQL is reasonable. It's just weird to issue two `UPDATE` SQL statements for the very same table twice in a row.